src/decode.rs

```diff
@@ -4179,626 +4179,465 @@ unsafe fn decode_b(
     0
 }
 
-unsafe extern "C" fn decode_sb(
-    t: *mut Dav1dTaskContext,
+unsafe fn decode_sb(
+    t: &mut Dav1dTaskContext,
     bl: BlockLevel,
     node: *const EdgeNode,
 ) -> libc::c_int {
-    let f: *const Dav1dFrameContext = (*t).f;
-    let ts: *mut Dav1dTileState = (*t).ts;
-    let hsz = 16 >> bl as libc::c_uint;
-    let have_h_split = ((*f).bw > (*t).bx + hsz) as libc::c_int;
-    let have_v_split = ((*f).bh > (*t).by + hsz) as libc::c_int;
-    if have_h_split == 0 && have_v_split == 0 {
-        if !((bl as libc::c_uint) < BL_8X8 as libc::c_int as libc::c_uint) {
-            unreachable!();
-        }
-        return decode_sb(
-            t,
-            (bl as libc::c_uint).wrapping_add(1 as libc::c_int as libc::c_uint) as BlockLevel,
-            (*(node as *const EdgeBranch)).split[0],
-        );
+    let f = &*t.f;
+    let ts = &mut *t.ts;
+    let hsz = 16 >> bl;
+    let have_h_split = f.bw > t.bx + hsz;
+    let have_v_split = f.bh > t.by + hsz;
+
+    if !have_h_split && !have_v_split {
+        assert!(bl < BL_8X8);
+        return decode_sb(t, bl + 1, (*(node as *const EdgeBranch)).split[0]);
     }
+
     let mut pc = &mut Default::default();
-    let mut bp: BlockPartition = PARTITION_NONE;
+    let mut bp;
     let mut ctx = 0;
     let mut bx8 = 0;
     let mut by8 = 0;
-    if (*t).frame_thread.pass != 2 as libc::c_int {
-        if 0 as libc::c_int != 0 && bl as libc::c_uint == BL_64X64 as libc::c_int as libc::c_uint {
-            printf(
-                b"poc=%d,y=%d,x=%d,bl=%d,r=%d\n\0" as *const u8 as *const libc::c_char,
-                (*(*f).frame_hdr).frame_offset,
-                (*t).by,
-                (*t).bx,
-                bl as libc::c_uint,
-                (*ts).msac.rng,
+    if t.frame_thread.pass != 2 {
+        if bl == BL_64X64 {
```

Comment:
`0 as libc::c_int != 0` is always false, so the condition should be `false && bl == BL_64X64`? This appears to have a significant impact on performance

Reply:
Ah, I think I must've misread that as `false ||` instead of `&&`.  I'll fix that.  Thanks for tracking it down!